product_merge/wizards/product_merge_wizard.py

```diff
@@ -0,0 +1,171 @@
+        store=True,
+        readonly=False,
+    )
+
+    @api.constrains("product_ids")
+    def _check_minimum_products(self):
+        """
+        Ensure that at least two products are selected for merging.
+        """
+        for wizard in self:
+            if len(wizard.product_ids) < 2:
+                raise ValidationError(
+                    _(
+                        "At least two products must be added to the wizard to perform a merge."
+                    )
+                )
+
+    @api.constrains("product_ids")
+    def _check_products_max_one_variant(self):
+        for wizard in self:
+            for product in wizard.product_ids:
+                if len(product.product_variant_ids) > 1:
+                    raise ValidationError(
+                        _(
+                            "All added products must have at most one variant. "
+                            "Product '%(product)s' has multiple variants.",
+                            product=product.name,
+                        )
+                    )
+
+    @api.constrains("product_ids")
+    def _check_product_types(self):
+        for wizard in self:
+            if len(wizard.product_ids) > 1:
+                types = wizard.product_ids.mapped("type")
+                if len(set(types)) > 1:
+                    raise ValidationError(
+                        _(
+                            "All products to merge must be of the same type "
+                            "(e.g., consumable, service, or storable)."
+                        )
+                    )
+
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        if self.env.context.get(
+            "active_model"
+        ) != "product.template" or not self.env.context.get("active_ids"):
+            return res
+        res["product_ids"] = [Command.set(self.env.context["active_ids"])]
+        return res
+
+    def action_merge_products(self):
+        self.ensure_one()
+        self.product_tmpl_id.with_context(product_merge=True).write(
+            {
+                "attribute_line_ids": [
+                    Command.create(
+                        {
+                            "attribute_id": attribute.id,
+                            "value_ids": [
+                                Command.link(value.id) for value in attribute.value_ids
+                            ],
+                        }
+                    )
+                    for attribute in self.attribute_ids
+                ],
+            }
+        )
+        other_templates = (
+            self.line_ids.product_id.product_tmpl_id - self.product_tmpl_id
+        )
+        for line in self.line_ids:
+            product_variant = line.product_id
+            self._update_supplier_info(product_variant)
+            self._move_variant_to_template(product_variant, line.attribute_value_ids)
+
+        other_templates.write({"active": False})
```

Comment:
@sbejaoui IMO something is missing. You should use a generic method to update into the database all the columns where the other_templates.ids are referenced.... The [`merge_records`](https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade_merge_records.py#L979) from the openupgrade lib could be a good candidate :thinking: